lib/wit/repo_entries.py

```diff
@@ -0,0 +1,116 @@
+            return Format.Manifest
+        return Format.unknown
+
+
+class RepoEntry:
+    def __init__(self, checkout_path, revision, remote_url, message=None):
+        # The path to checkout at within the workspace.
+        # JSON field name is 'name'.
+        self.checkout_path = checkout_path
+
+        # Desired revision that exists in the history of the below remote.
+        # JSON field name is 'commit'
+        self.revision = revision
+
+        # Url (or local fs path) for git to clone/fetch/push.
+        # JSON field name is 'source'
+        self.remote_url = remote_url
+
+        # A comment to leave in any serialized artifacts.
+        # Optional. JSON field name is '//'
+        self.message = message
+
+    def __repr__(self):
+        return str(self.__dict__)
+
+
+# OriginalEntry encodes the RepoEntry for both Lock and Manifest formats.
+class OriginalEntry():
+    @staticmethod
+    def to_dict(entry: RepoEntry) -> dict:
+        d = {
+            "name": entry.checkout_path,
+            "commit": entry.revision,
+            "source": entry.remote_url,
+        }
+        if entry.message and len(entry.message) > 0:
+            d["//"] = entry.message
+        return d
+
+    @staticmethod
+    def from_dict(data: dict) -> RepoEntry:
+        return RepoEntry(data["name"],
+                         data["commit"],
+                         data.get("source"),  # 'repo path' cli option needs this optional
+                         data.get("//"))  # optional
+
+
+# Utilities for List[RepoEntry]
+class RepoEntries:
+    @staticmethod
+    def write(path: Path, entries: List[RepoEntry]):
+        fmt = Format.from_path(path)
+        if fmt is Format.Manifest:
+            data = [OriginalEntry.to_dict(e) for e in entries]
+            json_data = json.dumps(data, sort_keys=True, indent=4) + "\n"
+        if fmt is Format.Lock:
+            ordered_data = OrderedDict((e.checkout_path, OriginalEntry.to_dict(e)) for e in entries)
+            json_data = json.dumps(ordered_data, sort_keys=True, indent=4) + "\n"
+        path.write_text(json_data)
+
+    @staticmethod
+    def read(path: Path) -> List[RepoEntry]:
+        text = path.read_text()
+        # 'parse' has to be decoupled from 'read' as sometimes
+        # we read files directly from the git object store rather
+        # than the filesystem
+        return RepoEntries.parse(text, path, "")
+
+    @staticmethod
+    def parse(text: str, path: Path, rev: str) -> List[RepoEntry]:
+        try:
+            fromtext = json.loads(text)
+        except json.JSONDecodeError as e:
+            print("Failed to parse json in {}:{}: {}".format(path, rev, e.msg))
+            sys.exit(1)
+
+        entries = []
+        fmt = Format.from_path(path)
+        for element in fromtext:
+            if fmt is Format.Manifest:
+                entry = element
+            if fmt is Format.Lock:
+                entry = fromtext[element]
```

Comment:
I realized that `fromtext` is either a list or a dict at the top level depending on which type of wit JSON file it is and that `for` loop iteration itself has different behavior depending on the file format, since with lists, you iterate over the list items, but with dicts, you iterate over the dictionary keys. This is kind of a surprising thing to read, even for someone who is aware of the different multiple file formats, and I was a bit confused about what was going on in the loop body because `fromtext` and `element` are different types depending on the input file.

Could I suggest that we hoist the `if fmt is Format.*` checks outside of the for loop so that it's clearer, and that you could iterate explicitly on the dictionary values for the dict case?

E.g.
```python
if fmt is Format.Manifest:
    for entry in fromtext:
        entries.append(OriginalEntry.from_dict(entry))
elif fmt is Format.Lock:
    for entry in fromtext.values():
        entries.append(OriginalEntry.from_dict(entry))
```

Reply:
yep will do